modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStorage.scala

```diff
@@ -150,41 +171,63 @@ object ConsensusStorage {
           (none[ConsensusState[Key, Artifact]], ()).some.pure[F]
         }.void
 
-      def findEvent(predicate: Event => Boolean): F[Option[Event]] =
-        for {
-          peerIds <- eventsR.keys
-          maybeFoundEvent <- peerIds.foldM(none[Event]) { (acc, peerId) =>
-            acc match {
-              case Some(foundEvent) => foundEvent.some.pure[F]
-              case None =>
-                eventsR(peerId).get.map {
-                  _.flatMap(events => events.map(_._2).find(predicate))
-                }
-            }
+      def containsTriggerEvent: F[Boolean] =
+        eventsR.keys.flatMap { keys =>
+          keys.existsM { peerId =>
+            eventsR(peerId).get
+              .map(_.flatMap(_.trigger).isDefined)
           }
-        } yield maybeFoundEvent
+        }
+
+      def addTriggerEvent(peerId: PeerId, peerEvent: (Ordinal, Event)): F[Unit] =
+        addEvents(peerId, List(peerEvent), updateTrigger = true)
 
       def addEvent(peerId: PeerId, peerEvent: (Ordinal, Event)): F[Unit] =
-        addEvents(Map(peerId -> List(peerEvent)))
+        addEvents(peerId, List(peerEvent), updateTrigger = false)
 
       def addEvents(events: Map[PeerId, List[(Ordinal, Event)]]): F[Unit] =
         events.toList.traverse {
           case (peerId, peerEvents) =>
-            eventsR(peerId).update { maybePeerEvents =>
-              (peerEvents ++ maybePeerEvents
-                .getOrElse(List.empty)).some
-            }
+            addEvents(peerId, peerEvents, updateTrigger = false)
         }.void
 
+      private def addEvents(peerId: PeerId, events: List[(Ordinal, Event)], updateTrigger: Boolean) =
+        eventsR(peerId).update { maybePeerEvents =>
+          maybePeerEvents
+            .getOrElse(PeerEvents.empty[Event])
+            .focus(_.events)
+            .modify(events ++ _)
+            .focus(_.trigger)
+            .modify { maybeCurrentTrigger =>
+              if (updateTrigger) {
+                val maybeNewTrigger = events.map(_._1).maximumOption
+
+                maybeCurrentTrigger match {
+                  case Some(currentTrigger) =>
+                    maybeNewTrigger.fold(currentTrigger)(o => Order[Ordinal].max(currentTrigger, o)).some
+                  case None => maybeNewTrigger
+                }
+              } else
+                maybeCurrentTrigger
```

Comment:
So you basically want to take `max` of 2 `Option[Ordinal]` or fallback to `maybeNewTrigger` or in worst case to `maybeCurrentTrigger`.

I guess then it can be simplified to:

```scala
(maybeCurrentTrigger, maybeNewTrigger)
   .mapN(Order[Ordinal].max(_, _))
   .orElse(maybeNewTrigger)
   .orElse(maybeCurrentTrigger)
```

eventually you can try with implicit conversion from Option to List and `maxOption` which uses `Ordering` implicitly

```scala
(maybeCurrentTrigger ++ maybeNewTrigger)
  .maxOption
  .orElse(maybeNewTrigger)
  .orElse(maybeCurrentTrigger)
```

Reply:
The first option you've suggested looks neat, I didn't think about that, fixing it.